### PR TITLE
vulkan: small mul_mat_vec optimizations

### PR DESCRIPTION
**Pull request by netrunnereve**
Here's a couple of small optimizations for `mul_mat_vec` that came from studying the generated GCN assembly. The biggest improvement came from moving out the delta if possible (8 additional multiplications per inner loop iteration add up fast).

Using the `dot()` function still generates FMA instructions but due to changes in the instruction ordering it's now running a little bit faster. I think on some newer cards there are dedicated dot product units which *may* have a better chance of being utilized by the compiler if we use this function.

I tested this on my RX 470, I just got it recently and I don't have access to the 570 anymore. Interestingly enough my 470 (and my W8100 as well) downclock pretty heavily when running `mul_mat_vec` as the hardware is hitting its power limit, it literally drops down from 1200 mhz to 750 during inference. This card has some VBIOS adjustments allowing the card to draw more power and run inference at around 1100 mhz.

Master:
| model                          |       size |     params | backend    | ngl | threads |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------------: | -------------------: |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        131.27 ± 0.17 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         25.19 ± 0.01 |
| llama 8B IQ4_NL - 4.5 bpw      |   4.35 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        127.40 ± 0.09 |
| llama 8B IQ4_NL - 4.5 bpw      |   4.35 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         23.60 ± 0.06 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        128.87 ± 0.13 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         16.53 ± 0.01 |
```
  MUL_MAT(type_a=q4_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   4260 runs -   273.43us/run - 117.44 MFLOP/run - 429.51 GFLOPS
  MUL_MAT(type_a=q4_1,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   4260 runs -   280.16us/run - 117.44 MFLOP/run - 419.20 GFLOPS
  MUL_MAT(type_a=q5_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   378.93us/run - 117.44 MFLOP/run - 309.92 GFLOPS
  MUL_MAT(type_a=q5_1,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   361.89us/run - 117.44 MFLOP/run - 324.52 GFLOPS
  MUL_MAT(type_a=q8_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   2556 runs -   441.92us/run - 117.44 MFLOP/run - 265.75 GFLOPS
  MUL_MAT(type_a=iq4_nl,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 4260 runs -   291.88us/run - 117.44 MFLOP/run - 402.36 GFLOPS
```

PR:
| model                          |       size |     params | backend    | ngl | threads |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------------: | -------------------: |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        131.09 ± 0.35 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         26.94 ± 0.07 |
| llama 8B IQ4_NL - 4.5 bpw      |   4.35 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        127.18 ± 0.18 |
| llama 8B IQ4_NL - 4.5 bpw      |   4.35 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         25.24 ± 0.07 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        129.33 ± 0.03 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         17.08 ± 0.00 |
```
  MUL_MAT(type_a=q4_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   4260 runs -   254.45 us/run - 117.44 MFLOP/run - 461.54 GFLOPS
  MUL_MAT(type_a=q4_1,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   4260 runs -   279.52 us/run - 117.44 MFLOP/run - 420.15 GFLOPS
  MUL_MAT(type_a=q5_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   350.07 us/run - 117.44 MFLOP/run - 335.48 GFLOPS
  MUL_MAT(type_a=q5_1,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   352.30 us/run - 117.44 MFLOP/run - 333.35 GFLOPS
  MUL_MAT(type_a=q8_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   2556 runs -   429.12 us/run - 117.44 MFLOP/run - 273.68 GFLOPS
  MUL_MAT(type_a=iq4_nl,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 4260 runs -   266.08 us/run - 117.44 MFLOP/run - 441.37 GFLOPS
```

By the way @0cc4m have you tried to implement the dequantization and matrix multiplication logic in F16 for the F16 shaders? The activations and deltas are already in F16 and it might not be necessary to do all the math in F32.

### CODE REVIEW

**Comment by jeffbolznv** [comment]
This is pretty close to neutral perfwise on my RTX 4070. Maybe a slight gain, but less than 0.5% and hard to quantify. In general, on recent hardware I expect mat-vec-mul to be pretty bandwidth constrained and optimizing math will have relatively small impact. That'll apply to switching to fp16 as well.

I'm concerned about having some dequantize functions need a separate multiply and some not, and by having different behavior between dequantize() and dequantize4() for the same format. If we're going to do this, I think it would be best if dequantize and dequantize4 have matching behavior, and all formats should have a "dequantizeScale" (or "dequantizeFactor"?) that returns the scale value and just have it return 1 for formats where it doesn't apply.

**Comment by ggerganov** [comment]
> By the way @0cc4m have you tried to implement the dequantization and matrix multiplication logic in F16 for the F16 shaders? The activations and deltas are already in F16 and it might not be necessary to do all the math in F32.

Seems like the product accumulators would have to remain in F32. I did some [experiments with Metal(https://github.com/ggerganov/llama.cpp/pull/10220) and apart from the PPL degradation when using full F16 math, there are some NaNs appearing in the CI which I haven't investigated yet in details, but suspect are because of this.

**Comment by 0cc4m** [comment]
> > By the way @0cc4m have you tried to implement the dequantization and matrix multiplication logic in F16 for the F16 shaders? The activations and deltas are already in F16 and it might not be necessary to do all the math in F32.
> 
> Seems like the product accumulators would have to remain in F32. I did some [experiments with Metal](https://github.com/ggerganov/llama.cpp/pull/10220) and apart from the PPL degradation when using full F16 math, there are some NaNs appearing in the CI which I haven't investigated yet in details, but suspect are because of this. For the Metal backend, I'm planning to put this behind an optional compile flag `GGML_METAL_F16` for starts.

Shouldn't this be covered by the precision choice in MUL_MAT?

@netrunnereve We added this precision switch for matrix matrix multiplication in #10206 and #10597, I guess this could also be done for matrix vector multiplication.

**Comment by ggerganov** [comment]
> Shouldn't this be covered by the precision choice in MUL_MAT?

It should - somehow I forgot about this feature.

**Comment by netrunnereve** [comment]
> This is pretty close to neutral perfwise on my RTX 4070. Maybe a slight gain, but less than 0.5% and hard to quantify. In general, on recent hardware I expect mat-vec-mul to be pretty bandwidth constrained and optimizing math will have relatively small impact. That'll apply to switching to fp16 as well.

I'm not sure how modern Nvidia architectures work but if it's like an Intel CPU where it rearranges instructions and processes several of them at once you might not notice the cost of the additional delta instruction. GCN runs them one at a time and that basically is a 5% difference in speed (the remaining 2% comes from my other changes). Personally I feel it's fine to make small improvements that help with low end systems provided that they don't complicate the code too much or slow down high end machines.

Also for my hardware I'm limited by GPU power and not memory bandwidth. If you're memory limited most optimizations here will have no effect since most of the time is spent waiting for the memory. The only thing I can think of that may help with inference speed in that case is to repack the quants so that the high and low bits are sequential rather than staggered, thus allowing us to read a continuous block of B memory. Alternatively we could do something like CUDA where we quantize the activations using Q8_0.

> I'm concerned about having some dequantize functions need a separate multiply and some not, and by having different behavior between dequantize() and dequantize4() for the same format. If we're going to do this, I think it would be best if dequantize and dequantize4 have matching behavior, and all formats should have a "dequantizeScale" (or "dequantizeFactor"?) that returns the scale value and just have it return 1 for formats where it doesn't apply.

Let me look into this. It's also worth noting that the regular `dequantize()` functions can be straight up removed at this point since we don't use them anymore.

> @netrunnereve We added this precision switch for matrix matrix multiplication in https://github.com/ggerganov/llama.cpp/pull/10206 and https://github.com/ggerganov/llama.cpp/pull/10597, I guess this could also be done for matrix vector multiplication.

This doesn't matter for my Polaris and Hawaii cards but pretty much everything newer runs F16 at twice the speed of F32. If the GPU isn't memory bound I think there would be an improvement in inference speed if we do the dot products in F16.

**Comment by jeffbolznv** [comment]
> Personally I feel it's fine to make small improvements that help with low end systems provided that they don't complicate the code too much or slow down high end machines.

Sure, I'm not objecting to the change! Just explaining why it's not showing as much benefit on NVIDIA.

File: ggml/src/ggml-vulkan/vulkan-shaders/mul_mat_vec.comp
Comment (0cc4m) [inline ggml/src/ggml-vulkan/vulkan-shaders/mul_mat_vec.comp:70]: I'd also prefer to handle this in a generic way, instead of with ifdefs.

**Review by 0cc4m — COMMENTED** [review]
I'm also not opposed to this change, of course. Let's just try to figure out a generic way to do this. If you need help let me know.

**Comment by netrunnereve** [comment]
> Let's just try to figure out a generic way to do this.

Okay this should do it. Performance is basically the same and the compiler is smart enough to optimize the branches appropriately for each quant type.

**Comment by netrunnereve** [comment]
I also tried adjusting the number of rows per workgroup and am seeing that I get the best results if the current values are doubled. If this only helps with GCN and slows down everyone else I'll add a check for this feature using the subgroup size (basically subgroup size 64 will run twice the number of rows per workgroup). Note that I made this change for FP16 as well but I'm only able to test with FP32.

Here are my new results:
| model                          |       size |     params | backend    | ngl | threads |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------------: | -------------------: |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        152.71 ± 0.38 |
| llama 8B Q4_0                  |   4.33 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         29.19 ± 0.04 |
| llama 8B IQ4_NL - 4.5 bpw      |   4.35 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        146.78 ± 0.44 |
| llama 8B IQ4_NL - 4.5 bpw      |   4.35 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         27.47 ± 0.22 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | Vulkan     | 100 |       8 |         pp512 |        149.42 ± 0.54 |
| llama 8B Q8_0                  |   7.95 GiB |     8.03 B | Vulkan     | 100 |       8 |         tg128 |         20.40 ± 0.01 |

```
  MUL_MAT(type_a=q4_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   5112 runs -   223.19 us run - 117.44 MFLOP/run - 526.19 GFLOPS
  MUL_MAT(type_a=q4_1,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   4260 runs -   266.12 us run - 117.44 MFLOP/run - 441.30 GFLOPS
  MUL_MAT(type_a=q5_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   320.84 us run - 117.44 MFLOP/run - 366.04 GFLOPS
  MUL_MAT(type_a=q5_1,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   345.09 us run - 117.44 MFLOP/run - 340.31 GFLOPS
  MUL_MAT(type_a=q8_0,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   343.47 us run - 117.44 MFLOP/run - 341.92 GFLOPS
  MUL_MAT(type_a=iq4_nl,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 5112 runs -   228.98 us run - 117.44 MFLOP/run - 512.89 GFLOPS
```

I'm also seeing that the FP16 pipeline for IQ4_NL on master is missing the NUM_ROWS setting after `device->subgroup_size`. Again I can't test this on my computer but it looks wrong.
https://github.com/ggerganov/llama.cpp/blob/3573fa8e7b7f0865638b52b4e9b4d2006f0558a2/ggml/src/ggml-vulkan/ggml-vulkan.cpp#L1813

**Comment by jeffbolznv** [comment]
Doubling the number of rows is pretty bad on NVIDIA:

```
NUM_ROWS=2
| chameleon 7B Q4_0              |   3.75 GiB |     7.01 B | Vulkan     | 1000 |         tg128 |         76.67  0.19 |
NUM_ROWS=4
| chameleon 7B Q4_0              |   3.75 GiB |     7.01 B | Vulkan     | 1000 |         tg128 |         71.42  1.04 |
```

I never really dug in to understand why it's so much slower. Making it depend on subgroup size seems ok, or just checking VendorID would also be fine.

**Comment by ggerganov** [comment]
FYI, I just added a Vulkan node to `ggml-ci` (GPU Tesla T4):

https://github.com/ggml-org/ci/tree/results/llama.cpp/ec/c93d0558fc3ecb8a5af69d2ece02fae4710ade/ggml-6-x86-vulkan-t4

It's currently failing due to some compile warnings, but these should be easy to resolve.

**Comment by netrunnereve** [comment]
> I'm also seeing that the FP16 pipeline for IQ4_NL on master is missing the NUM_ROWS setting after device->subgroup_size. Again I can't test this on my computer but it looks wrong.

I've fixed this one, please test this on a FP16 system before merging.

> I never really dug in to understand why it's so much slower. Making it depend on subgroup size seems ok, or just checking VendorID would also be fine.

I've made it check for both the vendor and subgroup size now so the changes should only apply to AMD GCN. Since you're memory bound I would expect it to run faster with more rows as the B vector is cached in registers so this is weird. Maybe Nvidia is confused by the memory access patterns as you're jumping around to read multiple rows, and the additional rows as well as the next set of A columns aren't prefetched properly.

> FYI, I just added a Vulkan node to ggml-ci (GPU Tesla T4):
>
> https://github.com/ggml-org/ci/tree/results/llama.cpp/ec/c93d0558fc3ecb8a5af69d2ece02fae4710ade/ggml-6-x86-vulkan-t4
>
> It's currently failing due to some compile warnings, but these should be easy to resolve.

Great! Our Vulkan backend is notorious for spitting out a lot of warnings so that will need to be dealt with... eventually.

Also I'm not sure who your provider is but you might be able to find something cheaper/older than the T4 as long as it supports Vulkan 1.3.

**Comment by ggerganov** [comment]
> Also I'm not sure who your provider is but you might be able to find something cheaper/older than the T4 as long as it supports Vulkan 1.3.

If you can suggest a cheaper instance on Azure cloud that is compatible with Vulkan - would be great. Currently using [Standard_NC4as_T4_v3](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/gpu-accelerated/ncast4v3-series?tabs=sizebasic)

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp
Comment (0cc4m) [inline ggml/src/ggml-vulkan/ggml-vulkan.cpp:1790]: This doesn't affect just GCN, but also RDNA. The subgroup_size check is basically redundant cause all AMD GPUs (which support Vulkan) have a subgroup_size of 64.

I see a slight regression with this row change on my 6800 XT. I don't have time right now, but I'll investigate further tomorrow or so.

If we do have to distinguish GCN from RDNA one way might be #10721, cause GCN has min and max subgroup sizes of 64, while RDNA has min 32 and max 64.

Comment (netrunnereve) [reply ggml/src/ggml-vulkan/ggml-vulkan.cpp:1790]: I've pulled in your https://github.com/ggerganov/llama.cpp/pull/10721 and used the min and max check to confirm that I'm using a GCN card. I think that's the only way to do it and it looks like there's nothing in Vulkan that tells you the precise architecture name of the graphics card.

In this case we'll need to merge this after https://github.com/ggerganov/llama.cpp/pull/10721.

**Comment by netrunnereve** [comment]
> If you can suggest a cheaper instance on Azure cloud that is compatible with Vulkan - would be great. Currently using [Standard_NC4as_T4_v3](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/gpu-accelerated/ncast4v3-series?tabs=sizebasic)

I took a quick look at the CI and since it runs a 7B model in FP16 you pretty much need the T4 as that's the cheapest Linux 16GB option on Azure. On the other hand you could probably save a bit on your cloud bill by switching the CUDA V100 for a T4 and the Metal M4 for a M1 or M2.

**Comment by jeffbolznv** [comment]
I think most of the current warnings in ggml-vulkan.cpp are my fault, so I'll fix them.

**Comment by ggerganov** [comment]
> Metal M4 for a M1 or M2.

The M4 is a self-hosted Mac Mini that I bought to run the CI.

**Comment by netrunnereve** [comment]
Hmm the CI is failing since it's using llvmpipe instead of the GPU.

**Comment by ggerganov** [comment]
> Hmm the CI is failing since it's using llvmpipe instead of the GPU.

It should be using the correct device now. But some tests are still failing:

https://github.com/ggml-org/ci/tree/results/llama.cpp/26/a8406ba9198eb6fdd8329fa717555b4f77f05f/ggml-6-x86-vulkan-t4

**Comment by jeffbolznv** [comment]
#10763 should fix the mat_mul failures. The im2col and rope failures are news to me.

**Comment by jeffbolznv** [comment]
I can reproduce the other failures on Turing, I'll debug them.

**Comment by jeffbolznv** [comment]
I filed https://github.com/ggerganov/llama.cpp/issues/10764 to track the Turing failures.

**Comment by netrunnereve** [comment]
Setting this as draft until https://github.com/ggerganov/llama.cpp/pull/10721 is merged.

**Comment by 0cc4m** [comment]
> Setting this as draft until #10721 is merged.

I'm sorry that it's taking a while. I'm still trying to figure out Intel and AMD. I think I'll merge a basic extension support version and then figure out how to deal with their specific issues.

**Comment by 0cc4m** [comment]
You probably have to rebase this PR so that it only contains your commits, once subgroup size control is merged. I think the conflict is from my branch, which I already rebased and fixed.

**Review by jeffbolznv — APPROVED** [review]
LGTM. I had previously tested on my system and perf was fine, I don't think I need to retest.

**Comment by netrunnereve** [comment]
Setting this back to draft as this updated logic is setting `device->subgroup_size_control` to 0 on my computer after the #10721 rebase. I'll need to dig more into this.
```
        device->subgroup_size_control = device->subgroup_size_control &&
                (subgroup_size_control_props.requiredSubgroupSizeStages & vk::ShaderStageFlagBits::eCompute) &&
                subgroup_size_control_features.subgroupSizeControl;
```

**Comment by netrunnereve** [comment]
Okay this actually works properly now and I'm getting the same performance as before. There's no harm in setting `subgroup_min_size` and `subgroup_max_size` even if `subgroup_size_control` is set to false.

**Comment by 0cc4m** [comment]
> Setting this back to draft as this updated logic is setting `device->subgroup_size_control` to 0 on my computer after the #10721 rebase. I'll need to dig more into this.
> ```
>         device->subgroup_size_control = device->subgroup_size_control &&
>                 (subgroup_size_control_props.requiredSubgroupSizeStages & vk::ShaderStageFlagBits::eCompute) &&
>                 subgroup_size_control_features.subgroupSizeControl;
> ```
> 

I'm surprised this doesn't go through on your system. Can you upload your vulkaninfo output? I thought this would be true on any modern driver.

**Comment by netrunnereve** [comment]
Unsurprisingly Intel's being its usual weird self again with NUM_ROWS=2. My 470 could probably beat it if I used a 7B model.

**Pull request merged**
